### PR TITLE
[WIP][IO] Do not process StreamerInfo Record at file opening if we processed an identical one already

### PR DESCRIPTION
**Pull request by dpiparo**

### CODE REVIEW

File: core/cont/inc/ROOT/RConcurrentHashColl.hxx [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+private:
+   mutable std::set<unsigned int> fHashes;
+   mutable ROOT::TRWSpinLock fRWLock;
+public:
+   RConcurrentUIntSet() = default;
+   /// If the hash is there, return false. Otherwise, insert the hash and return true;
+   bool Insert(unsigned int hash) const
+   {
+      {
+#ifdef R__USE_IMT
```
Comment (pcanal) [inline core/cont/inc/ROOT/RConcurrentHashColl.hxx]: Why the conditional here?  Wouldn't it better to have this container always thread safe and use it only when IMT is on?

**Comment by pcanal** [comment]
@phsft-bot build just slc6-i686/gcc49

File: core/cont/inc/ROOT/RConcurrentHashColl.hxx (hunk 1 above)
Comment (dpiparo) [reply core/cont/inc/ROOT/RConcurrentHashColl.hxx]: sure.

**Comment by pcanal** [comment]
> And run it, you need on my computer ~6 minutes w/o this patch and 11s with.

That is **very** nice :)  Thanks.

For completeness, what is the difference with and without the patch with a single thread (i.e. essentially cost of the Hashing vs cost of unstreaming).

**Comment by dpiparo** [comment]
Very good point @pcanal. So, with the patch, opening the same file 5k time 1s. W/o the patch 5.6s. So it is better also in the single threaded case. I think the reason is that it is much less work even if we go through the hashing.

**Comment by pcanal** [comment]
> , opening the same file 5k time 1s. W/o the patch 5.6s. 

Great.  That is what I was hoping.  This should even speed-up opening distinct but similar files (i.e. produced with the same set of StreamerInfo) and the cost for the case where the Hash is non-active (single files or series of files with distinct StreamerInfo set) seems to be negligeable.

Thanks!

**Comment by dpiparo** [comment]
Last performance measurement:
```.cpp
#include <ROOT/RDataFrame.hxx>

int main()
{
ROOT::EnableImplicitMT();
auto tdf = ROOT::RDataFrame("rechitntupler/hits", "DataForDanilo/ntuple_*.root");
tdf.Histo1D("run")->Draw();
return 0;
}
```
where `DataForDanilo/ntuple_*.root` is a set of 343 files with a size between 2.2MB to ~100KB. Without the patch, 8 cores, it takes 22s. With the patch 8s. And this is orthogonal to #2116.

**Comment by pcanal** [comment]
One last challenge though.  Since this is hash based, different StreamerInfo Sets will/can have the same hash value, what is the behavior (and what can we do about it) if the user ends up reading 2 such files in the same process?

**Pull request merged**